packages/lists/src/components/Viewport/Viewport.tsx

```diff
@@ -0,0 +1,157 @@
+import * as React from 'react';
+import { useState, useRef, useEffect } from 'react';
+import { useDebouncedTimeout } from './useDebouncedTimeout';
+import { usePreviousValue } from './usePreviousValue';
+
+export interface IViewportProps {
+  /**
+   * The height of the scroll container (in pixels).
+   */
+  height: number;
+
+  /**
+   * The width of the scroll container (in pixels).
+   */
+  width: number;
+
+  /**
+   * The child component is a function component which receives the scroll container's current scroll state as its only argument.
+   */
+  children: (viewportState: IViewportState) => JSX.Element[] | JSX.Element;
+}
+
+export type Vector2D<T> = [T, T];
+
+export enum Axis {
+  X = 0,
+  Y = 1
+}
+
+export enum ScrollDirection {
+  none,
+  backward,
+  forward
+}
+
+export interface IViewportState {
+  /**
+   * Whether the user is currently scrolling.
+   */
+  isScrolling: boolean;
+
+  /**
+   * Vector representing the distance scrolled from the top left corner, i.e. the point [0, 0], in the form
+   * [x, y] where x denotes the horizontal and y denotes the vertical scroll distance.
+   */
+  scrollDistance: Vector2D<number>;
+
+  /**
+   * Vector representing the current scroll direction in the form [x, y] where x denotes the horizontal and y
+   * denotes the vertical scroll direction.
+   */
+  scrollDirection: Vector2D<ScrollDirection>;
+}
+
+const SCROLL_DISTANCE_ORIGIN: Vector2D<Axis> = [0, 0];
+const NO_SCROLL_DIRECTION: Vector2D<ScrollDirection> = [ScrollDirection.none, ScrollDirection.none];
+
+const SCROLLING_RESET_TIMEOUT_IN_MILLISECONDS = 200;
+
+/**
+ * Determines the scroll direction based on the given current and previous scroll distance.
+ * @param scrollDistance The new scroll distance
+ * @param prevScrollDistance The previous scroll distance
+ */
+function getScrollDirection(scrollDistance: number, prevScrollDistance: number): ScrollDirection {
+  let scrollDirection = ScrollDirection.none;
+  if (scrollDistance > prevScrollDistance) {
+    scrollDirection = ScrollDirection.forward;
+  } else if (scrollDistance < prevScrollDistance) {
+    scrollDirection = ScrollDirection.backward;
+  }
+
+  return scrollDirection;
+}
+
+/**
+ * Viewport represents a scrollable container that maintains information about its current scroll state.
+ * The component takes a function component as its child component, using the current scroll state as its only argument.
+ */
+export const Viewport = React.memo(
+  (props: IViewportProps): JSX.Element => {
+    const { height, width, children } = props;
```

Comment:
I think things which receive children should not use `React.memo`. This children will pretty much always mutate on every render.

Reply:
The way I understood it in my reading is that the array or object of children always fail the referential equality check and thus break memoization. Could be wrong tho!

Reply:
As per our discussion offline, I will remove the `React.memo` and leave it up to the caller to decide whether to memoize the components against the params (since the caller can make sure that the `children` function doesn't mutate on every render).